### PR TITLE
Amd64/set trap context

### PR DESCRIPTION
**Pull request by tkreuzer**
This PR implements the following functions for x64:
- `RtlpCaptureNonVolatileContextPointers`, which walks the stack and captures the addresses of all non-volatile registers on the stack, when they have been saved first. This is needed to be able to fix up the non-volatile registers on a system call, which doesn't capture non-volatiles, but relies on them to be restored by the callees.
- `RtlSetUnwindContext`, which uses RtlpCaptureNonVolatileContextPointers to set the non-volatile registers in the the stack. They will be picked up, when returning back or unwinding, e.g. to the system call handler.
- `KiSetTrapContext`, which is an asm wrapper around RtlSetUnwindContext, which first stores an exception frame to assure that all non-volatile registers were put on the stack, then calls RtlSetUnwindContext to update their first saving positions on the stack and finally restore the exception frame to potentially load any updated registers, that haven't been saved elsewhere on the stack.

Finally it fixes `PspGetOrSetContextKernelRoutine` to use the above mechanism to change user mode context in the trap frame as a result of a call to `NtSetContextThread` / `PsSetContextThread`

I hope this sounds complicated enough to avoid too many irrelevant comments 😏

### CODE REVIEW

File: ntoskrnl/ps/amd64/psctx.c [hunk 1]
```diff
@@ -13,83 +13,65 @@
...
-//        TrapFrame = (PKTRAP_FRAME)((ULONG_PTR)Thread->InitialStack -
-//                                   ROUND_UP(sizeof(KTRAP_FRAME), KTRAP_FRAME_ALIGN) -
-//                                   sizeof(FX_SAVE_AREA));
+        /* Get the thread's base trap frame */
+        TrapFrame = KeGetTrapFrame(KeGetCurrentThread());
     }
 
     /* Check if it's a set or get */
-    if (Apc->SystemArgument1)
+    if (Apc->SystemArgument1 != 0)
```
Comment (HBelusca) [inline ntoskrnl/ps/amd64/psctx.c:60]:
```suggestion
    if (Apc->SystemArgument1 != NULL)
```

File: ntoskrnl/ke/amd64/trap.S [hunk 2]
```diff
@@ -1049,6 +1049,31 @@ KiConvertToGuiThreadFailed:
 
 ENDFUNC
 
+
+EXTERN KiSetTrapContextInternal:PROC
+
+/*
+ * VOID
+ * KiSetTrapContext (
+ *     _In_ PCONTEXT Context);
```
Comment (HBelusca) [inline ntoskrnl/ke/amd64/trap.S:1058]: Isn't the prototype
```
VOID
KiSetTrapContext(
    _Out_ PKTRAP_FRAME TrapFrame,
    _In_ PCONTEXT Context,
    _In_ KPROCESSOR_MODE RequestorMode);
```
instead?

File: sdk/lib/rtl/amd64/unwind.c [hunk 3]
```diff
@@ -687,3 +687,71 @@ RtlRaiseException(IN PEXCEPTION_RECORD ExceptionRecord)
     RtlRaiseStatus(Status);
 }
 
+VOID
+RtlCaptureNonVolatileContextPointers(
```
Comment (HBelusca) [inline sdk/lib/rtl/amd64/unwind.c:691]: Is this a name from official MS pdb symbols? If not, and if this used only internally, wouldn't it be better to use the `Rtlp` prefix instead?

File: ntoskrnl/ke/amd64/context.c [hunk 4]
```diff
@@ -292,3 +292,29 @@ KeTrapFrameToContext(IN PKTRAP_FRAME TrapFrame,
...
+    _In_ KPROCESSOR_MODE RequestorMode)
+{
+    ULONG64 TargetFrame;
+
+    /* Save the volatile register context in the trap frame */
+    KeContextToTrapFrame(Context,
+        NULL,
+        TrapFrame,
+        Context->ContextFlags,
+        RequestorMode);
```
Comment (HBelusca) [inline ntoskrnl/ke/amd64/context.c:313]:
```suggestion
    KeContextToTrapFrame(Context,
                         NULL,
                         TrapFrame,
                         Context->ContextFlags,
                         RequestorMode);
```

File: ntoskrnl/ke/amd64/context.c [hunk 5]
```diff
@@ -292,3 +292,29 @@ KeTrapFrameToContext(IN PKTRAP_FRAME TrapFrame,
...
+
+    /* The target frame is 0x80 bytes before the trap frame */
+    TargetFrame = (ULONG64)TrapFrame - 0x80;
```
Comment (HBelusca) [inline ntoskrnl/ke/amd64/context.c:316]: So, no #define or sthg for the "magic" 0x80 ?

File: ntoskrnl/ps/amd64/psctx.c (hunk 1 above)
Comment (tkreuzer) [reply ntoskrnl/ps/amd64/psctx.c:60]: Tn this case I actually prefer 0, since it's not a pointer, but an option stored in a pointer.

Comment (HBelusca) [reply ntoskrnl/ps/amd64/psctx.c:60]: Dunno... I see the type of the variable being PVOID 🤷

File: ntoskrnl/ke/amd64/context.c (hunk 5 above)
Comment (tkreuzer) [reply ntoskrnl/ke/amd64/context.c:316]: Fine, you'll get it.

File: ntoskrnl/ps/amd64/psctx.c (hunk 1 above)
Comment (tkreuzer) [reply ntoskrnl/ps/amd64/psctx.c:60]: Yeah, but I see what it's meaning is.

File: sdk/lib/rtl/amd64/unwind.c (hunk 3 above)
Comment (tkreuzer) [reply sdk/lib/rtl/amd64/unwind.c]: Changed.

File: ntoskrnl/ke/amd64/context.c (hunk 4 above)
Comment (tkreuzer) [reply ntoskrnl/ke/amd64/context.c]: kk

File: ntoskrnl/ke/amd64/context.c [hunk 6]
```diff
@@ -292,3 +292,29 @@ KeTrapFrameToContext(IN PKTRAP_FRAME TrapFrame,
     if (OldIrql < APC_LEVEL) KeLowerIrql(OldIrql);
 }
 
+VOID
+RtlSetUnwindContext(
+    _In_ PCONTEXT Context,
+    _In_ DWORD64 TargetFrame);
```
Comment (zefklop) [inline ntoskrnl/ke/amd64/context.c:298]: why DWORD64 ? PKTRAP_FRAME became a bad word ?

Comment (zefklop) [reply ntoskrnl/ke/amd64/context.c:298]: Errm, I now see it's a call site. Whaterver, then.

File: ntoskrnl/ps/amd64/psctx.c [hunk 7]
```diff
@@ -13,83 +13,65 @@
...
-{   
-    PAGED_CODE();
- 
-    /* Convert the trap frame to a context */
-    KeTrapFrameToContext(TrapFrame, NULL, Context);
-}
+KiSetTrapContext(
+    _Out_ PKTRAP_FRAME TrapFrame,
+    _In_ PCONTEXT Context,
+    _In_ KPROCESSOR_MODE RequestorMode);
```
Comment (zefklop) [inline ntoskrnl/ps/amd64/psctx.c:20]: can we have this in a header ?

File: sdk/lib/rtl/amd64/unwind.c [hunk 8]
```diff
@@ -687,3 +687,72 @@ RtlRaiseException(IN PEXCEPTION_RECORD ExceptionRecord)
...
+    KPROCESSOR_MODE Mode;
+
+    /* Zero out the nonvolatile context pointers */
+    RtlZeroMemory(NonvolatileContextPointers, sizeof(*NonvolatileContextPointers));
+
+    /* Capture the current context */
+    RtlCaptureContext(&Context);
+
+    /* Use the highest bit in RIP as mode flag */
+    Mode = Context.Rip >> 63;
```
Comment (zefklop) [inline sdk/lib/rtl/amd64/unwind.c:710]: RtlpGetMode() ? This looks fishy.

File: sdk/lib/rtl/amd64/unwind.c [hunk 9]
```diff
@@ -687,3 +687,72 @@ RtlRaiseException(IN PEXCEPTION_RECORD ExceptionRecord)
...
+                         ImageBase,
+                         Context.Rip,
+                         FunctionEntry,
+                         &Context,
+                         &HandlerData,
+                         &EstablisherFrame,
+                         NonvolatileContextPointers);
+
+        /* Continue until we reached the target frame or user mode */
+    } while ((EstablisherFrame < TargetFrame) &&
```
Comment (zefklop) [inline sdk/lib/rtl/amd64/unwind.c:729]: Are we sure that the frame will always go upwards ? What if there is some strange stuff done with setjmp/longjmp ?

Comment (tkreuzer) [reply sdk/lib/rtl/amd64/unwind.c:729]: longjmp is an unwind and actually goes through RtlUnwindEx (on Windows). Strange stuff with the stack is not allowed on x64.

Comment (zefklop) [reply sdk/lib/rtl/amd64/unwind.c:729]: OK.

File: sdk/lib/rtl/amd64/unwind.c (hunk 8 above)
Comment (tkreuzer) [reply sdk/lib/rtl/amd64/unwind.c]: I simplified it and added some ASSERTs instead.

File: ntoskrnl/ps/amd64/psctx.c (hunk 7 above)
Comment (tkreuzer) [reply ntoskrnl/ps/amd64/psctx.c]: Done

**Pull request merged**